File: c++/17140.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int arr[101][101];

bool compare(pair<int, int> a, pair<int, int> b) {
  if (a.second == b.second) {
    return a.first < b.first;
  }
  return a.second < b.second;
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  freopen("example.txt", "r", stdin);
  int r, c, k;
  cin >> r >> c >> k;

  int row = 3;
  int col = 3;

  for (int i = 1; i <= 3; i++) {
    for (int j = 1; j <= 3; j++) {
      cin >> arr[i][j];
    }
  }

  int t = 0;

  while (t <= 100) {
    if (arr[r][c] == k) {
      cout << t;
      return 0;
    }

    if (row >= col) {
      for (int i = 1; i <= row; i++) {
        int temp[101] = {0};
        vector<pair<int, int>> v;
        int maxNum = 0;
        for (int j = 1; j <= col; j++) {
          int num = arr[i][j];
          temp[num]++;
        }
        for (int j = 1; j <= 100; j++) {
          if (temp[j] != 0) {
            v.push_back({j, temp[j]});
          }
        }
        sort(v.begin(), v.end(), compare);
        for (int j = 1; j <= col; j++) arr[i][j] = 0;
        if (v.size() * 2 > col) col = v.size() * 2;
        int idx = 1;
        for (int j = 0; j < v.size(); j++) {
          arr[i][idx++] = v[j].first;
          arr[i][idx++] = v[j].second;
        }
      }
    } else {
      for (int i = 1; i <= col; i++) {
        int temp[101] = {0};
        vector<pair<int, int>> v;
        int maxNum = 0;
        for (int j = 1; j <= row; j++) {
          int num = arr[j][i];
          temp[num]++;
        }
        for (int j = 1; j <= 100; j++) {
          if (temp[j] != 0) {
            v.push_back({j, temp[j]});
          }
        }
        sort(v.begin(), v.end(), compare);
        for (int j = 1; j <= row; j++) arr[j][i] = 0;
        if (v.size() * 2 > row) row = v.size() * 2;
        int idx = 1;
        for (int j = 0; j < v.size(); j++) {
          arr[idx++][i] = v[j].first;
          arr[idx++][i] = v[j].second;
        }
      }
    }
    t++;
  }

  cout << -1;

  return 0;
}

File: js/17140.js
const fs = require('fs');

const inputs = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split('\n')
  .map((e) => e.split(' ').map(Number));

const [r, c, k] = inputs.shift();

const arr = Array(101)
  .fill()
  .map((_, i) =>
    Array(101)
      .fill()
      .map((_, j) => {
        if (1 <= i && i <= 3 && 1 <= j && j <= 3) return inputs[i - 1][j - 1];
        return 0;
      })
  );

let t = 0;
let row = 3;
let col = 3;
let answer = -1;

while (t <= 100) {
  if (arr[r][c] === k) {
    answer = t;
    break;
  }

  if (row >= col) R();
  else C();

  t++;
}

console.log(answer);

function R() {
  for (let i = 1; i <= row; i++) {
    const temp = Array(101).fill(0);
    let maxNum = 0;
    for (let j = 1; j <= col; j++) {
      const num = arr[i][j];
      maxNum = Math.max(maxNum, num);
      temp[num]++;
    }
    const next = [];
    for (let j = 1; j <= maxNum; j++) {
      if (temp[j] !== 0) {
        next.push([j, temp[j]]);
      }
    }
    next.sort((a, b) => {
      if (a[1] === b[1]) return a[0] - b[0];
      return a[1] - b[1];
    });
    const flated = next.flat();

    if (flated.length > col) col = flated.length;

    for (let j = 1; j <= col; j++) {
      if (flated.length >= j) arr[i][j] = flated[j - 1];
      else arr[i][j] = 0;
    }
  }
}

function C() {
  for (let i = 1; i <= col; i++) {
    const temp = Array(101).fill(0);
    let maxNum = 0;
    for (let j = 1; j <= row; j++) {
      const num = arr[j][i];
      maxNum = Math.max(maxNum, num);
      temp[num]++;
    }
    const next = [];
    for (let j = 1; j <= maxNum; j++) {
      if (temp[j] !== 0) {
        next.push([j, temp[j]]);
      }
    }

    next.sort((a, b) => {
      if (a[1] === b[1]) return a[0] - b[0];
      return a[1] - b[1];
    });
    const flated = next.flat();

    if (flated.length > row) row = flated.length;

    for (let j = 1; j <= row; j++) {
      if (flated.length >= j) arr[j][i] = flated[j - 1];
      else arr[j][i] = 0;
    }
  }
}
